rcl/src/rcl/event.c

```diff
@@ -0,0 +1,182 @@
+// Copyright 2019 Open Source Robotics Foundation, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifdef __cplusplus
+extern "C"
+{
+#endif
+
+#include "rcl/event.h"
+
+#include <stdio.h>
+
+#include "rcl/error_handling.h"
+#include "rcl/expand_topic_name.h"
+#include "rcl/remap.h"
+#include "rcutils/logging_macros.h"
+#include "rmw/error_handling.h"
+#include "rmw/validate_full_topic_name.h"
+#include "rmw/event.h"
+
+#include "./common.h"
+#include "./publisher_impl.h"
+#include "./subscription_impl.h"
+
+
+typedef struct rcl_event_impl_t
+{
+  rmw_event_t rmw_handle;
+  rcl_allocator_t allocator;
+} rcl_event_impl_t;
+
+
+rcl_event_t
+rcl_get_zero_initialized_event()
+{
+  static rcl_event_t null_event = {0};
+  return null_event;
+}
+
+rcl_ret_t
+rcl_publisher_event_init(
+  rcl_event_t * event,
+  const rcl_publisher_t * publisher,
+  const rcl_publisher_event_type_t event_type)
+{
+  rcl_ret_t ret = RCL_RET_OK;
+
+  // Check publisher and allocator first, so allocator can be used with errors.
+  RCL_CHECK_ARGUMENT_FOR_NULL(publisher, RCL_RET_INVALID_ARGUMENT);
+  rcl_allocator_t * allocator = &publisher->impl->options.allocator;
+  RCL_CHECK_ALLOCATOR_WITH_MSG(allocator, "invalid allocator", return RCL_RET_INVALID_ARGUMENT);
+
+  // Allocate space for the implementation struct.
+  event->impl = (rcl_event_impl_t *) allocator->allocate(
+    sizeof(rcl_event_impl_t), allocator->state);
+  RCL_CHECK_FOR_NULL_WITH_MSG(
+    event->impl, "allocating memory failed", ret = RCL_RET_BAD_ALLOC; return ret);
+
+  event->impl->rmw_handle = rmw_get_zero_initialized_event();
+  event->impl->allocator = *allocator;
+
+  rmw_event_type_t rmw_event_type = RMW_EVENT_INVALID;
+  switch (event_type) {
+    case RCL_PUBLISHER_OFFERED_DEADLINE_MISSED:
+      rmw_event_type = RMW_EVENT_OFFERED_DEADLINE_MISSED;
+      break;
+    case RCL_PUBLISHER_LIVELINESS_LOST:
+      rmw_event_type = RMW_EVENT_LIVELINESS_LOST;
+      break;
+    default:
+      return RCL_RET_INVALID_ARGUMENT;
```

Comment:
You should set the error message in this case to something specific about the event type being unknown.

Reply:
Good point, added